### PR TITLE
Dm 1405 epics 7

### PR DESCRIPTION
**Pull request by SkyToGround**
### Issue

* Closes DM-1405

### Description of work

This is an update of the forwarder to use EPICS v7. In Order for this to work, several changes had to be made. These include:
* Copying of *caRepeater* for channel access to work.
* Addition of code to make it possible for the EPICS library to find the *caRepeater* application.
* Fixing the code for instantiating channel access support.
* Change the code such that only one channel access channel provider is ever instantiated in order to work around a bug in the EPICS library.
* Fix minor namespace changes.

I also made some other minor improvements.

### CODE REVIEW

File: src/CAPathSetup.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+#ifdef _WIN32
+#include <direct.h>
+#include <stdlib.h>
+#else
+#include <unistd.h>
+#endif
+
+bool fileExists(std::string const &FullPath) {
+  std::fstream InFile(FullPath);
+  return InFile.good();
```
Comment (matthew-d-jones) [inline src/CAPathSetup.cpp]: I'm still looking into the failure in the system tests, but hit this issue first.
For some reason I didn't have write permissions over caRepeater so fileExists always fails:
```
-r-xr-xr-x  1 matt matt    16736 May 21 15:05 caRepeater
```
Using `fopen` would avoid this problem:
```suggestion
  if (FILE *File = fopen(FullPath.c_str(), "r")) {
    fclose(File);
    return true;
  } else {
    return false;
  }
```
and change `#include <fstream>` to `#include <stdexcept>`.

Comment (matthew-d-jones) [reply src/CAPathSetup.cpp]: In kafkacow we resorted to boost: https://github.com/ess-dmsc/kafkacow/blob/master/src/GetSchemaPath.cpp but if we can make your approach work consistently then it is probably the lesser of the evils.
When the mac nodes are finally updated we can use `std::filesystem`: https://github.com/ess-dmsc/kafkacow/pull/109

**Comment by matthew-d-jones** [comment]
@SkyToGround 

I see the same on my machine (using --local-build system-test option) as in the logs on Jenkins (https://jenkins.esss.dk/dm/job/ess-dmsc/job/forward-epics-to-kafka/job/PR-328/26/artifact/system-tests/logs/forwarder_tests_forwarding_2020-05-20.log). It is failing to connect to the PVs.
Could be a CaRepeater config issue, I'll read up a bit and see if I can figure anything out, but you probably know more than me about it.

Have you tested it with a local build somehow? If you have EPICS env variables set on your system then that could be causing the difference.

File: src/CAPathSetup.cpp (hunk 1 above)
Comment (SkyToGround) [reply src/CAPathSetup.cpp]: Yes, I am looking forward to being able to use `std::filesystem`. In this particular case, I am confident that using `std::ifstream` is a better option than using `fopen(..., "r")` however. Please test with the updated code.

**Comment by matthew-d-jones** [comment]
Disregard my suggestion about CaRepeater, the behaviour looks the same for PVA so it is unlikely that is the cause.

**Pull request closed without merging**